src/Agda2Hs/Compile/RuntimeCheckUtils.hs

```diff
@@ -0,0 +1,108 @@
+module Agda2Hs.Compile.RuntimeCheckUtils where
+
+import Agda.Syntax.Common
+import Agda.Syntax.Common.Pretty (prettyShow)
+import Agda.Syntax.Internal
+import Agda.TypeChecking.InstanceArguments (findInstance)
+import Agda.TypeChecking.MetaVars (newInstanceMeta, newLevelMeta)
+import Agda.TypeChecking.Monad (addContext, liftTCM)
+import Agda.TypeChecking.Pretty (prettyTCM)
+import Agda.TypeChecking.Reduce (instantiate)
+import Agda.TypeChecking.Telescope (splitTelescopeAt)
+import Agda2Hs.Compile.Term (compileTerm)
+import Agda2Hs.Compile.Types (C)
+import Agda2Hs.Compile.Utils
+import Agda2Hs.HsUtils
+import Control.Monad.Except (catchError)
+import Data.List (partition)
+import Data.Maybe (catMaybes, isJust)
+import qualified Language.Haskell.Exts as Hs
+
+-- Checks if an argument is erased
+checkErased :: Dom (ArgName, a) -> Bool
+checkErased t = case getQuantity t of
+  Quantity0 _ -> True
+  _ -> False
+
+-- Convenience function to prepend signatures if the check exists
+maybeAppend :: [a] -> Maybe a -> Maybe [a]
+maybeAppend sig chk = (sig ++) . (: []) <$> chk
+
+-- Replace unnamed arguments with a${i}
+-- NB this does not handle nested but neither does the encapsulating logic yet
+nameArgs :: [(Dom (ArgName, a), Int)] -> [String]
+nameArgs = map $ \(d, i) -> let name = fst $ unDom d in if name == "_" then "a" ++ show i else name
+
+-- Turn a type into its Dec version
+decify :: Type -> C Type
+decify t = do
+  dec <- resolveStringName "Haskell.Extra.Dec.Dec"
+  level <- liftTCM newLevelMeta
+  let vArg = defaultArg
+      hArg = setHiding Hidden . vArg
+  return $ t {unEl = Def dec $ map Apply [hArg $ Level level, vArg $ unEl t]}
+
+-- Failably find instances for decidable terms
+findDecInstances :: Type -> C (Maybe Term)
+findDecInstances t =
+  liftTCM $
+    do
+      (m, v) <- newInstanceMeta "" t
+      findInstance m Nothing
+      Just <$> instantiate v
+      `catchError` (\_ -> return Nothing)
+
+-- Create expression to be put in the guard
+createGuardExp :: Type -> C (Maybe (Hs.Exp ()))
+createGuardExp t = do
+  dec <- decify t
+  sth <- findDecInstances dec
+  -- XXX functor something?
+  findDecInstances dec >>= \case
+    Just r -> Just <$> compileTerm dec r
+    Nothing -> return Nothing
```

Comment:
This really feels to me like an fmap should be possible, but I don't know how.

Reply:
something like `(fmap $ compileTerm dec r) =<< findDecInstances dec`?

Reply:
`copileTerm dec r <$> sequence (findDecInstanes dec)`, I think?

Reply:
Nope, `sequence (findDecInstances dec)` expects a `C (C Term)` (idk why exactly), but is a `Maybe (C Term)`

Reply:
"expects" as in "has the type"?

Reply:
Just with `sth <- sequence $ findDecInstances dec` (to avoid confusion with the extra error messages that appear with your suggested fmap):
```
    • Couldn't match type ‘Maybe’ with ‘C’
      Expected: C (C Term)
        Actual: Maybe (C Term)
    • In a stmt of a 'do' block: sth <- sequence $ findDecInstances dec
```
(where I replaced
```
                  Control.Monad.Trans.RWS.CPS.RWST
                  Agda2Hs.Compile.Types.CompileEnv
                  Agda2Hs.Compile.Types.CompileOutput
                  Agda2Hs.Compile.Types.CompileState
                  Agda.TypeChecking.Monad.Base.TCM
```
with `C` for brevity)

Reply:
But how can I not be in a `C` block when I need to call `compileTerm :: Type -> Term -> C (Exp ())`?

Reply:
the result of `copileTerm dec <$> sequence (findDecInstanes dec)` is a `Maybe (C Term)`, if you want to bind it in a `C` do-block I guess you can do another `sequence` on the result?

Alternatively you could try doing
```
r1 <- findDecInstances dec
traverse (compileTerm dec) r1
```

Or maybe `mapMM` ([hackage](https://hackage.haskell.org/package/Agda-2.6.4.3/docs/Agda-Utils-Monad.html#v:mapMM))is what you're after here, to get something like `mapMM (compileTerm dec) (findDecInstances dec)`

Reply:
`traverse`, I was looking for `traverse`. Can even just `findDecInstances dec >>= traverse (compileTerm dec)`. Thank you! Update coming.

Reply:
https://impurepics.com/posts/2020-10-03-always-traverse.html